### ISSUE TITLE
Is Proper Ratio Support on the Roadmap?

### ISSUE THREAD

**Issue by BSlug** [udem-dlteam/ribbit#67]
Currently, I believe e.g. (/ 9 2) evaluates to 4. I’ve read that float support is unlikely to happen for portability reasons, but ratios could offer a middleground.

**Comment by l4haie** [comment]
We’re actively working towards supporting floats in ribbit, hopefully this will be done by the end of summer!

**Comment by BSlug** [comment]
Will there be a way to easily exclude floats from the numerical stack, for systems such as microcontrollers where the float routines would increase the binary size considerably?

**Comment by leo-ard** [comment]
yes ! This will be done through the feature system of ribbit. To have proper floats, you will need to enable the "float" feature

**Comment by raviqqe** [comment]
I've recently implemented 62-bit floating-point numbers which is called `flonum`s internally in Ruby in [my Scheme implementation](https://github.com/raviqqe/nonbox/blob/d5e7e6900abb2dc9acd95b1875e2e3c1097c7c77/src/f62.rs) following Ribbit Scheme's architecture. That might be also interesting for Ribbit Scheme because it appears to be faster for boxing/unboxing integers and cons pointers than traditional NaN boxing on 64-bit floats.

- https://github.com/raviqqe/stak/pull/2323
- https://github.com/raviqqe/nonbox/blob/d5e7e6900abb2dc9acd95b1875e2e3c1097c7c77/src/f62.rs

@l4haie was also working on [the float/bignum implementation](https://github.com/raviqqe/nonbox/blob/d5e7e6900abb2dc9acd95b1875e2e3c1097c7c77/src/f62.rs). But the PR seems to be forsaken now...

**Comment by leo-ard** [comment]
Hey @raviqqe! Thanks for sharing your implementation of float62. I read through it and as I understand it you embed the tags inside the 2 highest bits of the matissa of a classical IEEE float64 (with the rotation operator). Then for operations you need to unbox/rebox them each time and reuse the rust operations to make it work (without any allocations, which is nice). Is this what is used in Ruby? It seems similar to self-tagging (https://arxiv.org/pdf/2411.16544)

@l4haie was indeed working on a float/bignum implementation. Most of the work here was for the number->string/string->number functions and bignums. We wanted to have a solution that is potable to all 16 hosts, thus implementing this logic in scheme. We choose to only keep operations on floats as host-specific, because emulating them in scheme would be a huge performance bottleneck. I still have plans to reintroduce the work that @l4haie did last summer this summer and might tweek it so that Ribbit can avoid allocations for floats (which would be very nice!)

**Comment by raviqqe** [comment]
@leo-ard Thank you for sharing the fantastic research work of Ribbit Scheme too :)

> Is this what is used in Ruby?

Yes, I think this is still used in Ruby. In [this `rb_num2dbl` function](https://github.com/ruby/ruby/blob/c218862d3c664b4afff5acce55d7a6eb13779809/object.c#L3827), there are two function calls of [`rb_float_flonum_value`](https://github.com/ruby/ruby/blob/c218862d3c664b4afff5acce55d7a6eb13779809/object.c#L3834C20-L3834C41) for tagged-value floats and [`rb_float_noflonum_value`](https://github.com/ruby/ruby/blob/c218862d3c664b4afff5acce55d7a6eb13779809/object.c#L3843C20-L3843C43) for boxed floats. The name of `flonum` is a bit confusing, but it specifically indicates the unboxed 62-bit floats.

> It seems similar to self-tagging

Thank you for sharing the info too. It looks similar to the 62-bit floats in Ruby while self-tagging doesn't need the extra rotation for most integers.

> We choose to only keep operations on floats as host-specific, because emulating them in scheme would be a huge performance bottleneck.

I see that integer/flonum/bignum checks are also implemented completely in Scheme in the PR. Are you planning to continue with the solution? While I'm implementing the 64-bit floating-point numbers, I've noticed that number type checks can degrade performance even when they are implemented in the host language of Rust if I don't escape the checks into a separate path.

**Comment by raviqqe** [comment]
The check-aware primitives look really smart! It's great that fallbacks are done only by `if`  and `call` instructions against primitives although I'm not sure how big their performance impact is.

In Stak Scheme, I implemented the `let`-like syntaxes using the reference implementations in the R7RS specification document which are like the ones in normal lambda calculus. ;) Maybe I should revert the primitive-based `let` implementation back...
